### ISSUE TITLE
When Dart gets weak references, improve ScalableImageCache/ScalableImageSource

### ISSUE THREAD

**Issue by zathras** [zathras/jovial_svg#10]
Since `ScalableImage` instances are immutable, it would be good to canonicalize the `ScalableImage` instances that come out of `ScalableImageSource`, and are cached in `ScalableImageCache`.  This, of course, requires weak references, which (as of this writing) Dart does not have.  It seems this is mostly because JavaScript hasn't quite gotten there yet.

If soft references (caching weak references) are ever added, a global soft cache would be something to consider, to avoid the potential for repeated loading/decoding of a `ScalableImageWidget` loaded with the `fromSISource` factory.

Because these things aren't currently possible, it is **strongly** recommended that users of jovial_svg manage `ScalableImage` objects as part of their application state.  Perhaps the most straightforward way to do this is to maintain SI instances directly.  Alternately, another way of doing this with `fromSISource` is to have a `ScalableImageCache` of an appropriate size whose lifetime is appropriate for the given assets.  This could be as simple as a global cache created by the application and held in a static data member.

See also:
https://github.com/dart-lang/language/issues/1847
https://github.com/dart-lang/language/pull/1853
https://github.com/dart-lang/language/blob/master/working/1847%20-%20FinalizationRegistry/proposal.md

**Comment by HosamHasanRamadan** [comment]
Dart 2.17 will have `weak reference` see:

https://github.com/dart-lang/sdk/blob/main/CHANGELOG.md#dartcore

**Comment by zathras** [comment]
> Dart 2.17 will have `weak reference` see:
> 
> https://github.com/dart-lang/sdk/blob/main/CHANGELOG.md#dartcore

Great news, thanks!  I did a bit of digging, and it looks like they're getting mature in JavaScript - Firefox has supported weak refs since July '21, it seems.  As of Jan. 2022, only IE and Opera don't support them.  Still, it makes sense for Dart to be a little conservative around timing.

**Comment by zathras** [comment]
Hmmm...  The semantic guarantees around Flutter's weak reference class are really weak (no pun intended).  Specifically, a `Finalizer` instance would need to be attached to every weakly-referenced `ScalableImage` in order to clear out the key in the cache.  As of Flutter 3.3, there is **no** guarantee that finalizers will ever be invoked, even if memory is running out.  I assume that they usually are called on most platforms most of the time, and I'd guess that the web runtime is the problematic one.

Still, if I did this optimization, I think the right thing to do would be to make it default to **off**, so that the client would be forced to make the hard decision of which platforms aren't safe.  And, anyone who understands this stuff well enough to make a good decision on that question would almost certainly be able to implement their own weak cache easily enough if that actually provided value.  Which is unlikely, since it's usually better to just manage lifetimes of expensive stuff in such a way that caching is irrelevant, or where that's not possible, manage caching in a way that's more appropriate to the application at hand, not using weak references at all.  Like by using `ScalableImageCache`.

So, I guess I've talked myself into believing that, given the lack of any semantic guarantees around finalizers, this style of caching is only really useful in platform-dependent code that's limited to platforms where finalizer behavior is more predictable.
